.github/workflows/tests.yml

```diff
@@ -213,26 +222,50 @@ jobs:
             core.setOutput('test-bot-formulae-args', test_bot_formulae_args.join(" "))
             core.setOutput('test-bot-dependents-args', test_bot_dependents_args.join(" "))
 
+  setup_runners:
+    needs: [formulae_detect, setup_tests]
+    if: >
+      github.event_name == 'pull_request' &&
+      fromJson(needs.setup_tests.outputs.syntax-only) == false
+    runs-on: ubuntu-22.04
+    container:
+      image: ghcr.io/homebrew/ubuntu22.04:master
+    outputs:
+      runners: ${{steps.determine-runners.outputs.runners}}
+      runners_present: ${{steps.determine-runners.outputs.runners_present}}
+      dependent_runners: ${{steps.determine-dependent-runners.outputs.runners}}
+      dependent_runners_present: ${{steps.determine-dependent-runners.outputs.runners_present}}
+    env:
+      HOMEBREW_LINUX_RUNNER: ${{needs.setup_tests.outputs.linux-runner}}
+      HOMEBREW_LINUX_CONTAINER_IMAGE: ghcr.io/homebrew/ubuntu22.04:master
+      HOMEBREW_LINUX_CONTAINER_OPTIONS: --user=linuxbrew -e GITHUB_ACTIONS_HOMEBREW_SELF_HOSTED
+      HOMEBREW_LINUX_WORKDIR: /github/home
+      HOMEBREW_LINUX_TIMEOUT: 4320
+      HOMEBREW_LINUX_CLEANUP: ${{needs.setup_tests.outputs.cleanup-linux-runner}}
+      TESTING_FORMULAE: ${{needs.formulae_detect.outputs.testing_formulae}}
+      DELETED_FORMULAE: ${{needs.formulae_detect.outputs.deleted_formulae}}
+    steps:
+      - name: Set up Homebrew
+        id: set-up-homebrew
+        uses: Homebrew/actions/setup-homebrew@master
+
+      - name: Determine runners to use for tests job
+        id: determine-runners
+        run: brew determine-test-runners "$TESTING_FORMULAE" "$DELETED_FORMULAE"
+
+      - name: Determine runners to use for test_deps job
+        id: determine-dependent-runners
+        run: brew determine-test-runners --dependents "$TESTING_FORMULAE"
+
   tests:
-    needs: setup_tests
-    if: ${{github.event_name == 'pull_request' && fromJson(needs.setup_tests.outputs.syntax-only) == false}}
+    needs: [tap_syntax, setup_tests, setup_runners]
+    if: >
+      github.event_name == 'pull_request' &&
+      fromJson(needs.setup_tests.outputs.syntax-only) == false &&
+      fromJson(needs.setup_runners.outputs.runners_present)
     strategy:
       matrix:
-        include:
-          - runner: "13-arm64-${{github.run_id}}-${{github.run_attempt}}"
```

Comment:
This approach won't work if we ever want to make a 🟢 `tests` build a requirement because you'll need to have the named build actually run in that case.

Reply:
Yup, I mention this at https://github.com/Homebrew/homebrew-core/pull/127236#issuecomment-1492993923.

> If we want to make `tests` a required status check at some point, it should be straightforward to modify this to run a job on an `ubuntu-latest` runner that does nothing.

Reply:
@carlocab Gotcha. I think I'd like that to be part of the implementation before it's merged, if that's alright 🙇🏻

Reply:
Implemented.

Reply:
You can see the latest run deploy an `ubuntu-latest` runner that does basically nothing. (i.e. a `mkdir` here, an `rm -rvf` there...)

Reply:
Thanks @carlocab 🙇🏻

Reply:
@carlocab Actually, I don't think this quite works as expected/needed:
- the name `tests (ubuntu-latest, true)` seems to have changed. Not opposed to a change here but I think getting rid of that `true` would be preferable.
- this doesn't seem to be creating equivalently named jobs for e.g. macOS which means we won't be able to prevent e.g. someone merging a build when macOS is failing.

Reply:
> * the name `tests (ubuntu-latest, true)` seems to have changed. Not opposed to a change here but I think getting rid of that `true` would be preferable.

This is a GitHub thing -- the `true` is from the second part of the `matrix` job. Not quite sure how to get GitHub to stop showing parts of the build matrix, but I'll look into it.

> * this doesn't seem to be creating equivalently named jobs for e.g. macOS which means we won't be able to prevent e.g. someone merging a build when macOS is failing.

Why would it create a macOS job? There are no formulae to test. If there were formulae to be tested here that supported macOS, then macOS jobs would be deployed.

Reply:
> If we don't create some sort of macOS job or job-named-the-same-as-macOS-job: we can never require the macOS job be 🟢 before merge.

I'm confused here -- isn't the point to require 🟢 status for `tests` before merging? If we do that, then failing formulae builds cannot be merged, even after this change.

> How much time does a change like this PR save and how many formulae does this affect?

At least 5 minutes per PR for formulae with no dependents (which is most of them). Plus the extra time saved from waiting less time in the queue because they're waiting for runners spending 4~ minutes doing a no-op job, which happens to pretty much every formula. So somewhere between 4-9 minutes per PR, I think. That's also not counting the time we waste re-running jobs because they timed out because of the extra time they spent on no-op stuff.

Reply:
Looks like the way to do required checks for `matrix` jobs is to add a check for jobs after them that depend on the `matrix` jobs: https://brunoscheufler.com/blog/2022-04-09-the-required-github-status-check-that-wasnt

Reply:
> status for the `tests` job before merging

There is no `tests` job. When you use a `matrix`: each part of the matrix is a separate job. You need to require success of each of these with the same name. If we ever want to prevent people from merging a e.g. failing macOS 13 ARM job: we always need to generate a job with that name and it always needs to be 🟢.

> At least 5 minutes per PR for formulae with no dependents (which is most of them).

To be clear: I think it's reasonable to skip the `test_deps` job in these situations: I just don't think we ever want to skip any `tests` jobs. Eventually: `CI-syntax-only` should will need to go away.

Reply:
> Looks like the way to do required checks for `matrix` jobs is to add a check for jobs after them that depend on the `matrix` jobs

Yes, if you must have a dynamic metric. I don't really like that, though; it means if the matrix logic ends up not working as you expect: you don't actually have the protections you think you do.

Reply:
> I don't really like that, though; it means if the matrix logic ends up not working as you expect: you don't actually have the protections you think you do.

I am very confident about the logic here for the `tests` job. (On the other hand, I am less confident about it for `test_deps`, but edge cases should be relatively rare. This is because `HOMEBREW_SIMULATE_MACOS_ON_LINUX` doesn't let you specify which macOS version you want to simulate.)

In any case: we might need to make a post-`tests` job as the required check even without this. We still have some `tests` jobs that take 24+ hours, and when that happens, failures are inevitable because we can't stop `GITHUB_TOKEN` from expiring.

> Eventually: `CI-syntax-only` should will need to go away.

Being able to skip OSes in `tests` would go a long way towards this, I think.

Reply:
> I am very confident about the logic here for the tests job.

Ok. As long as much of this logic as possible is pushed to Homebrew/brew and/or Homebrew/homebrew-test-bot and has tests over there: I'm fine with this 👍🏻 

> failures are inevitable because we can't stop `GITHUB_TOKEN` from expiring.

I wouldn't say inevitable here. This seems exactly the sort of thing that we need to fix at some point.

> Being able to skip OSes in tests would go a long way towards this, I think.

Agreed.

Reply:
Is the fake "macOS 13-arm64" runner in the latest CI run here more like what you're looking for?

It's fake because it's actually `ubuntu-latest`.

Reply:
I think your `conclusion` approach would be fine here. It has just occurred to me that even the current format (`CI / tests (13-arm64-4626220015-1) (pull_request)`) we can't actually require because the test name changes every time 😬